runtime/src/iree/hal/drivers/hip/cleanup_thread.c

```diff
@@ -0,0 +1,160 @@
+  has_request |= thread->do_exit;
+  iree_slim_mutex_unlock(&thread->mutex);
+  return has_request;
+}
+
+static int iree_hal_hip_cleanup_thread_main(void* param) {
+  iree_hal_hip_cleanup_thread_t* thread = (iree_hal_hip_cleanup_thread_t*)param;
+  bool exit = false;
+  while (true) {
+    iree_notification_await(&thread->notification,
+                            &iree_hal_hip_cleanup_thread_has_request, thread,
+                            iree_infinite_timeout());
+
+    iree_slim_mutex_lock(&thread->mutex);
+    exit |= thread->do_exit;
+    iree_status_t status = thread->failure_status;
+    while (!iree_hal_hip_callback_queue_empty(&thread->queue)) {
+      iree_hal_hip_cleanup_thread_callback_t callback =
+          iree_hal_hip_callback_queue_at(&thread->queue, 0);
+      iree_hal_hip_callback_queue_pop_front(&thread->queue, 1);
+      iree_slim_mutex_unlock(&thread->mutex);
+
+      if (iree_status_is_ok(status)) {
+        status = IREE_HIP_RESULT_TO_STATUS(
+            thread->symbols,
+            hipEventSynchronize(iree_hal_hip_event_handle(callback.event)));
+      }
+
+      status = iree_status_join(
+          status,
+          callback.callback(callback.user_data, callback.event, status));
+      iree_slim_mutex_lock(&thread->mutex);
+      if (!iree_status_is_ok(status)) {
+        thread->failure_status = status;
+      }
+      if (!iree_status_is_ok(thread->failure_status)) {
+        status = iree_status_clone(thread->failure_status);
+      }
+    }
+    iree_slim_mutex_unlock(&thread->mutex);
+
+    if (!iree_status_is_ok(status) || exit) {
+      break;
+    }
+  }
+  return 0;
+}
+
+iree_status_t iree_hal_hip_cleanup_thread_initialize(
+    const iree_hal_hip_dynamic_symbols_t* symbols,
+    iree_allocator_t host_allocator,
+    iree_hal_hip_cleanup_thread_t** out_thread) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_hal_hip_cleanup_thread_t* thread = NULL;
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0,
+      iree_allocator_malloc(host_allocator, sizeof(*thread), (void**)&thread));
+
+  thread->symbols = symbols;
+  thread->do_exit = false;
+  iree_slim_mutex_initialize(&thread->mutex);
+  iree_hal_hip_callback_queue_initialize(host_allocator, &thread->queue);
+  thread->failure_status = iree_ok_status();
+  thread->host_allocator = host_allocator;
+  iree_notification_initialize(&thread->notification);
+
+  iree_thread_create_params_t params;
+  memset(&params, 0x00, sizeof(params));
+  params.name = iree_make_cstring_view("iree-hal-hip-cleanup");
+  iree_status_t status =
+      iree_thread_create((iree_thread_entry_t)iree_hal_hip_cleanup_thread_main,
+                         thread, params, host_allocator, &thread->thread);
+  if (!iree_status_is_ok(status)) {
+    iree_hal_hip_callback_queue_deinitialize(&thread->queue);
+    iree_slim_mutex_deinitialize(&thread->mutex);
+    iree_allocator_free(host_allocator, thread);
+  }
+  IREE_TRACE_ZONE_END(z0);
+  out_thread[0] = thread;
```

Comment:
only use array syntax on arrays (since we don't really have it in the language we have to be consistent so it's clear what the intent is)
```suggestion
  *out_thread = thread;
```